File: python/백준/다이나믹 프로그래밍/9084 동전.py
import sys

input = sys.stdin.readline

T = int(input())


for _ in range(T):

    N = int(input())

    coins = list(map(int, input().split()))

    target = int(input())
    d = [0] * (target+1)
    d[0] = 1

    for coin in coins:
        for money in range(coin, target + 1):
            d[money] += d[money - coin]

    print(d[target])

File: cpp/백준/다이나믹 프로그래밍/9084 동전.cpp
#include <iostream>
#include <vector>
using namespace std;

int main() {
    ios_base::sync_with_stdio(false);
    cin.tie(nullptr);

    int T;
    cin >> T;

    while (T--) {
        int N;
        cin >> N;

        vector<int> coins(N);
        for (int i = 0; i < N; ++i) cin >> coins[i];

        int target;
        cin >> target;

        vector<int> d(target + 1, 0);
        d[0] = 1;

        for (int coin : coins) {
            for (int money = coin; money <= target; ++money) {
                d[money] += d[money - coin];
            }
        }

        cout << d[target] << "\n";
    }

    return 0;
}
